### PR TITLE
Nsxv plugin Enhancements

### PR DESCRIPTION
**Pull request by rarcotvmw**
The header options for the nsx policy will support securitygroup securitygroup-id as the target options. These are optional. If these target options are given, all the rules in the section will have appliedTo as the given Security Group.
If the values are not given, the appliedTo will be default to DISTRIBUTED_FIREWALL

### CODE REVIEW

File: lib/nsxv.py [hunk 1]
```diff
@@ -326,14 +329,25 @@ def __str__(self):
     for s in services:
       service = '%s%s' % (service, s)
 
+    #applied To
```
Comment (ankenyr) [inline lib/nsxv.py]: Is this a comment or something else? There should be a space after the # if this is a comment and the capitalization is incorrect for a sentence. Please fix.

File: lib/nsxv.py [hunk 2]
```diff
@@ -326,14 +329,25 @@ def __str__(self):
...
+    appliedToList = ''
```
Comment (ankenyr) [inline lib/nsxv.py]: This is an invalid name for a variable. Please see the python style guide we try to maintain.
https://google.github.io/styleguide/pyguide.html?showone=Naming#Naming

You will want to fix this for appliedToList and appliedTo

File: lib/nsxv.py [hunk 3]
```diff
@@ -326,14 +329,25 @@ def __str__(self):
...
+    if self.applied_to:
+      appliedToList = '<appliedToList>'
+      appliedTo = '%s%s%s' % (_XML_TABLE.get('appliedToStart'),
+                              self.applied_to,
+                              _XML_TABLE.get('appliedToEnd'))
+      appliedToList = '%s%s' %(appliedToList, appliedTo)
+      appliedToList = '%s%s' %(appliedToList, '</appliedToList>')
+
```
Comment (ankenyr) [inline lib/nsxv.py]: too many blank lines here, please remove one. Two blank lines come before a class, one is allowed elsewhere

File: lib/nsxv.py [hunk 4]
```diff
@@ -326,14 +329,25 @@ def __str__(self):
...
     # action
     action = '%s%s%s' % (_XML_TABLE.get('actionStart'),
                          _ACTION_TABLE.get(str(self.term.action[0])),
                          _XML_TABLE.get('actionEnd'))
 
     ret_lines = []
-    ret_lines.append('<rule logged="%s"> %s %s %s %s %s %s </rule>' %
-                     (log, name, action, sources, destinations, service, notes))
+    ret_lines.append('<rule logged="%s"> %s %s %s %s %s %s %s </rule>' %
+                     (log, name, action, sources, destinations, service, appliedToList, notes))
```
Comment (ankenyr) [inline lib/nsxv.py]: This line is too long, we maintain a 80 character line limit.

File: lib/nsxv.py [hunk 5]
```diff
@@ -503,30 +505,83 @@ def _TranslatePolicy(self, pol, exp_info):
...
+            new_terms.append(Term(inet_term, filter_type, applied_to, 4))
           else:
             inet6_term = self.FixHighPorts(term, 'inet6')
             if not inet6_term:
               continue
-            new_terms.append(Term(inet6_term, filter_type, 6))
+            new_terms.append(Term(inet6_term, filter_type, applied_to, 6))
 
-      self.nsxv_policies.append((header, filter_name, [filter_type],
+      self.nsxv_policies.append((header, filter_name, [filter_type], section_id,
```
Comment (ankenyr) [inline lib/nsxv.py]: Why are we returning section_id? This breaks from the norm.

File: lib/nsxv.py [hunk 6]
```diff
@@ -503,30 +505,83 @@ def _TranslatePolicy(self, pol, exp_info):
...
                                  new_terms))
 
+  def _ParseFilterOptions(self, filter_options):
+    """Parses the target in header for filter type, section_id and applied_to.
```
Comment (ankenyr) [inline lib/nsxv.py]: Please put a new line after the first line in a docstring

File: lib/nsxv.py [hunk 7]
```diff
@@ -503,30 +505,83 @@ def _TranslatePolicy(self, pol, exp_info):
...
+       target:: nsxv mixed 12345 securitygroup securitygroupId
```
Comment (ankenyr) [inline lib/nsxv.py]: Please see other doc strings or the following link on how to construct a docstring
https://google.github.io/styleguide/pyguide.html?showone=Comments#Comments

You need to create the sections for arguments, returns and errors and correctly format them

File: lib/nsxv.py [hunk 8]
```diff
@@ -503,30 +505,83 @@ def _TranslatePolicy(self, pol, exp_info):
...
+       SecurityGroup securitygroupId is also optional
+
+       The element followed by 'securitygroup' is the securitygroupId.
+    """
+    filter_options_dict = {}
+
+    # check for filter type
+    if not filter_options:
+      raise UnsupportedNsxvAccessListError(
+        'Filter type is not provided for %s'  % (self._PLATFORM))
```
Comment (ankenyr) [inline lib/nsxv.py]: You need to add two spaces here to correctly indent this line

File: lib/nsxv.py [hunk 9]
```diff
@@ -503,30 +505,83 @@ def _TranslatePolicy(self, pol, exp_info):
...
+    filter_type = filter_options[0]
+
+    # a mixed filter outputs both ipv4 and ipv6 acls in the same output file
+    good_filters = ['inet', 'inet6', 'mixed']
+
+    # check if filter type is renderable
+    if filter_type not in good_filters:
+      raise UnsupportedNsxvAccessListError(
+          'Access list type %s not supported by %s (good types: %s)' % (
+             filter_type, self._PLATFORM, str(good_filters)))
```
Comment (ankenyr) [inline lib/nsxv.py]: Add one space on this line.

File: lib/nsxv.py [hunk 10]
```diff
@@ -503,30 +505,83 @@ def _TranslatePolicy(self, pol, exp_info):
...
+    section_id = 0
+    applied_to = None
+    filter_opt_len = len(filter_options)
+
+    if filter_opt_len > 1:
+      for index in range(1, filter_opt_len):
+        if index == 1 and filter_options[1] != 'securitygroup':
+          section_id = filter_options[1]
+          continue
+        if (filter_options[index] == 'securitygroup'):
```
Comment (ankenyr) [inline lib/nsxv.py]: You do not need parens here.

File: lib/nsxv.py [hunk 11]
```diff
@@ -503,30 +505,83 @@ def _TranslatePolicy(self, pol, exp_info):
...
+          if (index + 1 <= filter_opt_len - 1):
```
Comment (ankenyr) [inline lib/nsxv.py]: Nor here.

File: lib/nsxv.py [hunk 12]
```diff
@@ -503,30 +505,83 @@ def _TranslatePolicy(self, pol, exp_info):
...
+            applied_to = filter_options[index + 1]
+            break
+          else:
+            raise UnsupportedNsxvAccessListError('Security Group Id is not provided for %s' % (self._PLATFORM))
```
Comment (ankenyr) [inline lib/nsxv.py]: This line is too long also

File: lib/nsxv.py [hunk 13]
```diff
@@ -573,4 +622,4 @@ def __str__(self):
       target = target_header + target
       target.append('%s' % (_XML_TABLE.get('sectionEnd')))
       target.append('\n')
-    return '\n'.join(target)
+    return '\n'.join(target)
```
Comment (ankenyr) [inline lib/nsxv.py]: Please always have a newline at the end of the file

File: tests/lib/nsxv_test.py [hunk 14]
```diff
@@ -337,23 +385,45 @@ def testStrForinet6(self):
...
                                                   nacaddr.IP('172.16.0.0/12'),
                                                   nacaddr.IP('192.168.0.0/16')])
     self.naming.GetServiceByProto.return_value = ['123']
 
-    exp_info = 2
     pol = policy.ParsePolicy(INET_FILTER, self.naming, False)
-    translate_pol = nsxv.Nsxv(pol, exp_info)
+    translate_pol = nsxv.Nsxv(pol, EXP_INFO)
+    nsxv_policies = translate_pol.nsxv_policies
+    for (header, filter_name, filter_list, section_id, terms) in nsxv_policies:
```
Comment (ankenyr) [inline tests/lib/nsxv_test.py]: Why did you change the underscore to a named variable? _ is used when a variable is unused. Please take a look at the following to understand how to mark a variable as unused.
https://google.github.io/styleguide/pyguide.html?showone=Lint#Lint

You should also think about why section_id is returned here. It seems weird

File: tests/lib/nsxv_test.py [hunk 15]
```diff
@@ -337,23 +385,45 @@ def testStrForinet6(self):
...
+    self.naming.GetNetAddr('INTERNAL').AndReturn([nacaddr.IP('10.0.0.0/8'),
+                                                  nacaddr.IP('172.16.0.0/12'),
+                                                  nacaddr.IP('192.168.0.0/16')])
+    self.naming.GetServiceByProto.return_value = ['123']
+
+    pol = policy.ParsePolicy(INET_FILTER_WITH_ESTABLISHED, self.naming, False)
+    translate_pol = nsxv.Nsxv(pol, EXP_INFO)
     nsxv_policies = translate_pol.nsxv_policies
-    for (_, filter_name, filter_list, terms) in nsxv_policies:
+    for (header, filter_name, filter_list, section_id, terms) in nsxv_policies:
```
Comment (ankenyr) [inline tests/lib/nsxv_test.py]: Same here, please adopt our style guide and use underscores to mark variables that are unused.

File: tests/lib/nsxv_test.py [hunk 16]
```diff
@@ -450,6 +517,44 @@ def testNsxvStr(self):
...
+    if 'Id' not in comment:
+      self.fail('Id missing in xml comment in test_nsxv_str()')
+    if 'Date' not in comment:
+      self.fail('Date missing in xml comment in test_nsxv_str()')
+    if 'Revision' not in comment:
+      self.fail('Revision missing in xml comment in test_nsxv_str()')
+
+    root = ET.fromstring(sections[1])
+    # check section name
+    section_name = {'id': '1010', 'name': 'Sample NSXV filter with Security Group'}
```
Comment (ankenyr) [inline tests/lib/nsxv_test.py]: Line too long.

File: tests/lib/nsxv_test.py [hunk 17]
```diff
@@ -475,6 +580,38 @@ def testBuildWarningTokens(self):
     self.assertEquals(st, SUPPORTED_TOKENS)
     self.assertEquals(sst, SUPPORTED_SUB_TOKENS)
 
+  def testParseFilterOptionsCase1(self):
+    pol = nsxv.Nsxv(policy.ParsePolicy(HEADER_WITH_SECTIONID + TERM, self.naming, False), EXP_INFO)
```
Comment (ankenyr) [inline tests/lib/nsxv_test.py]: Line too long.

File: tests/lib/nsxv_test.py [hunk 18]
```diff
@@ -475,6 +580,38 @@ def testBuildWarningTokens(self):
...
+    for (header, filter_name, filter_list, section_id, terms) in pol.nsxv_policies:
```
Comment (ankenyr) [inline tests/lib/nsxv_test.py]: Again unused stuff, and too long, please fix.

File: tests/lib/nsxv_test.py [hunk 19]
```diff
@@ -475,6 +580,38 @@ def testBuildWarningTokens(self):
...
+      filter_options = header.FilterOptions(_PLATFORM)
+      filter_options_dict = pol._ParseFilterOptions(filter_options)
+      self.assertEquals(filter_options_dict['filter_type'],'inet')
```
Comment (ankenyr) [inline tests/lib/nsxv_test.py]: A space is needed after the comma.

File: tests/lib/nsxv_test.py [hunk 20]
```diff
@@ -475,6 +580,38 @@ def testBuildWarningTokens(self):
...
+      self.assertEquals(filter_options_dict['section_id'],'1009')
```
Comment (ankenyr) [inline tests/lib/nsxv_test.py]: Same here.

File: tests/lib/nsxv_test.py [hunk 21]
```diff
@@ -475,6 +580,38 @@ def testBuildWarningTokens(self):
...
+      self.assertEquals(filter_options_dict['applied_to'], None)
+
+  def testParseFilterOptionsCase2(self):
+    pol = nsxv.Nsxv(policy.ParsePolicy(HEADER_WITH_SECURITYGROUP + INET6_TERM, self.naming, False), EXP_INFO)
```
Comment (ankenyr) [inline tests/lib/nsxv_test.py]: Line too long.

File: tests/lib/nsxv_test.py [hunk 22]
```diff
@@ -475,6 +580,38 @@ def testBuildWarningTokens(self):
...
+    for (header, filter_name, filter_list, section_id, terms) in pol.nsxv_policies:
```
Comment (ankenyr) [inline tests/lib/nsxv_test.py]: More unused and line too long

File: tests/lib/nsxv_test.py [hunk 23]
```diff
@@ -475,6 +580,38 @@ def testBuildWarningTokens(self):
...
+      filter_options = header.FilterOptions(_PLATFORM)
+      filter_options_dict = pol._ParseFilterOptions(filter_options)
+      self.assertEquals(filter_options_dict['filter_type'],'inet6')
```
Comment (ankenyr) [inline tests/lib/nsxv_test.py]: Space after comma.

File: tests/lib/nsxv_test.py [hunk 24]
```diff
@@ -475,6 +580,38 @@ def testBuildWarningTokens(self):
...
+      self.assertEquals(filter_options_dict['section_id'],0)
```
Comment (ankenyr) [inline tests/lib/nsxv_test.py]: Space after comma.

File: tests/lib/nsxv_test.py [hunk 25]
```diff
@@ -475,6 +580,38 @@ def testBuildWarningTokens(self):
...
+
+  def testBadHeaderCase2(self):
+    pol = policy.ParsePolicy(BAD_HEADER_1 + TERM, self.naming, False)
+    self.assertRaises(nsxv.UnsupportedNsxvAccessListError,
+                      nsxv.Nsxv, pol, EXP_INFO)
+
+  def testBadHeaderCase3(self):
+    pol = policy.ParsePolicy(BAD_HEADER_2 + INET6_TERM, self.naming, False)
+    self.assertRaises(nsxv.UnsupportedNsxvAccessListError,
+                      nsxv.Nsxv, pol, EXP_INFO )
```
Comment (ankenyr) [inline tests/lib/nsxv_test.py]: Remove the space at the end here

File: tests/lib/nsxv_test.py [hunk 26]
```diff
@@ -475,6 +580,38 @@ def testBuildWarningTokens(self):
...
 
 if __name__ == '__main__':
-  unittest.main()
+  unittest.main()
```
Comment (ankenyr) [inline tests/lib/nsxv_test.py]: Please add a newline at the end of this file.

File: lib/nsxv.py (hunk 5 above)
Comment (rarcotvmw) [inline lib/nsxv.py]: We need the section id to populate the section element of the xml

**Comment by rarcotvmw** [comment]
Addressed review comments

File: lib/nsxv.py (hunk 5 above)
Comment (ankenyr) [inline lib/nsxv.py]: This can be part of the header and in fact is already.
(Pdb) print header.FilterOptions('nsxv')
[u'inet', u'1010', u'securitygroup', u'securitygroup-Id']

Comment (rarcotvmw) [inline lib/nsxv.py]: Will make the change.

File: lib/nsxv.py [hunk 27]
```diff
@@ -326,14 +329,25 @@ def __str__(self):
     for s in services:
       service = '%s%s' % (service, s)
 
+    # applied_to
+    appliedto_list = ''
```
Comment (ankenyr) [inline lib/nsxv.py]: You want to have underscores between words in a variable
applied_to_list

File: lib/nsxv.py [hunk 28]
```diff
@@ -326,14 +329,25 @@ def __str__(self):
...
+    if self.applied_to:
+      appliedto_list = '<appliedToList>'
+      appliedTo = '%s%s%s' % (_XML_TABLE.get('appliedToStart'),
```
Comment (ankenyr) [inline lib/nsxv.py]: Looks like this one was missed. Variable names should be lower case and underscore separated.

**Comment by rarcotvmw** [comment]
Addressed all the review comments. 
Please let me know if any other changes are required.

Thank you

File: lib/nsxv.py [hunk 29]
```diff
@@ -326,14 +329,25 @@ def __str__(self):
     for s in services:
       service = '%s%s' % (service, s)
 
+    # applied_to
+    applied_to_list = ''
+    if self.applied_to:
+      applied_to_list = '<appliedToList>'
+      applied_to_element = '%s%s%s' % (_XML_TABLE.get('appliedToStart'),
+                              self.applied_to,
```
Comment (ankenyr) [inline lib/nsxv.py]: Thanks for fixing this one but the indentation is wrong, You need to align it by adding 9 spaces.

File: lib/nsxv.py [hunk 30]
```diff
@@ -326,14 +329,25 @@ def __str__(self):
...
+                              _XML_TABLE.get('appliedToEnd'))
```
Comment (ankenyr) [inline lib/nsxv.py]: Same here please

Comment (rarcotvmw) [inline lib/nsxv.py]: Done. Can you please let me know which editor do you use? I have used PyCharm. Also is there any template for the code formatting? Next time I can make sure these indentation is correct.

Thank you

File: lib/nsxv.py [hunk 31]
```diff
@@ -503,30 +505,85 @@ def _TranslatePolicy(self, pol, exp_info):
...
+      A dictionary that contains fields necessary to create the firewall
+      rule.
+
+    Raises:
+      UnsupportedNsxvAccessListError: Raised when we're give a non named access
+      list.
+    """
+    # check for filter type
+    if not filter_options:
+      raise UnsupportedNsxvAccessListError(
```
Comment (ankenyr) [inline lib/nsxv.py]: Looks like there are no tests to hit this line

File: lib/nsxv.py [hunk 32]
```diff
@@ -503,30 +505,85 @@ def _TranslatePolicy(self, pol, exp_info):
...
+          'Filter type is not provided for %s'  % (self._PLATFORM))
+
+    if len(filter_options) > 4:
```
Comment (ankenyr) [inline lib/nsxv.py]: Nor this one.

File: tests/lib/nsxv_test.py [hunk 33]
```diff
@@ -224,6 +268,10 @@
     },
 }
 
+# Print a info message when a term is set to expire in that many weeks.
+# This is normally passed from command line.
+EXP_INFO = 2
+_PLATFORM = 'nsxv'
```
Comment (rarcotvmw) [inline tests/lib/nsxv_test.py]: Done

File: tests/lib/nsxv_test.py [hunk 34]
```diff
@@ -367,9 +413,32 @@ def testTranslatePolicyWithEstablished(self):
...
+      self.assertEqual(filter_list, ['inet'])
+      self.assertEqual(len(terms), 1)
+
+      self.assertNotIn('<sourcePort>123</sourcePort><destinationPort>123',
+                       str(terms[0]))
+
+    self.naming.GetServiceByProto.assert_has_calls(
+        [mock.call('NTP', 'udp')] * 2)
+
+  def testTranslatePolicyWithEstablished(self):
```
Comment (rarcotvmw) [inline tests/lib/nsxv_test.py]: Looks like this got added again when trying to resolve merge conflict. Removed this one.

File: tests/lib/nsxv_test.py [hunk 35]
```diff
@@ -475,6 +580,40 @@ def testBuildWarningTokens(self):
...
+
+  def testParseFilterOptionsCase2(self):
+    pol = nsxv.Nsxv(policy.ParsePolicy(HEADER_WITH_SECURITYGROUP + INET6_TERM,
+                                       self.naming, False), EXP_INFO)
+    for (header, _, _, _) in pol.nsxv_policies:
+      filter_options = header.FilterOptions(_PLATFORM)
+      pol._ParseFilterOptions(filter_options)
+      self.assertEquals(nsxv.Nsxv._FILTER_OPTIONS_DICT['filter_type'], 'inet6')
+      self.assertEquals(nsxv.Nsxv._FILTER_OPTIONS_DICT['section_id'], 0)
+      self.assertEquals(nsxv.Nsxv._FILTER_OPTIONS_DICT['applied_to'], 'securitygroup-Id1')
```
Comment (rarcotvmw) [inline tests/lib/nsxv_test.py]: Done

File: lib/nsxv.py (hunk 31 above)
Comment (rarcotvmw) [inline lib/nsxv.py]: Added a test case.

File: lib/nsxv.py (hunk 32 above)
Comment (rarcotvmw) [inline lib/nsxv.py]: Added a test case.
